aten/src/ATen/native/cuda/Activation.cu

```diff
@@ -243,6 +247,160 @@ std::tuple<Tensor, Tensor> prelu_backward_cuda(const Tensor& grad_out_, const Te
   return std::tuple<Tensor, Tensor>{input_grad, weight_grad};
 }
 
+// -----------------------------------
+// rrelu
+// -----------------------------------
+template<typename scalar_t>
+inline scalar_t __device__ curand_uniform_type(curandStatePhilox4_32_10_t *state);
+
+template <>
+inline THHalf __device__ curand_uniform_type<THHalf>(curandStatePhilox4_32_10_t *state) {
+  auto rand = curand_uniform4(state);
+  return ScalarConvert<float, THHalf>::to(rand.x);
+}
+
+template <>
+inline float __device__ curand_uniform_type<float>(curandStatePhilox4_32_10_t *state) {
+  auto rand = curand_uniform4(state);
+  return rand.x;
+}
+
+template <>
+inline double __device__ curand_uniform_type<double>(curandStatePhilox4_32_10_t *state) {
+  auto rand = curand_uniform2_double(state);
+  return rand.x;
+}
+
+template <typename scalar_t>
+__global__ void rrelu_with_noise_cuda_kernel(
+    int64_t n,
+    PhiloxCudaState philox_args,
+    scalar_t* output,
+    scalar_t* input,
+    scalar_t* noise,
+    double a,
+    double b) {
+  auto seeds = at::cuda::philox::unpack(philox_args);
+  int idx = blockIdx.x * blockDim.x + threadIdx.x;
+  curandStatePhilox4_32_10_t state;
+  curand_init(std::get<0>(seeds),
+              idx,
+              std::get<1>(seeds),
+              &state);
+
+  CUDA_KERNEL_LOOP(i, n)
+  {
+    if (input[i] <= 0)
```

Comment:
to avoid warp divergence, you should generate randoms for every input, and then only diverge on fast operations like computing output and noise